subsys/net/lib/ftp_client/src/ftp_client.c

```diff
@@ -0,0 +1,793 @@
+	if (ret != FTP_CODE_200) {
+		close(client.sock);
+		return ret;
+	}
+
+	LOG_DBG("FTP opened");
+	return ret;
+}
+
+/**@brief API to login FTP connection
+ */
+int ftp_login(const char *username, const char *password)
+{
+	int ret;
+	int keepalive_time = CONFIG_FTP_KEEPALIVE_TIME;
+
+	/* send username */
+	sprintf(ctrl_buf, CMD_USER, username);
+	ret = do_ftp_send_ctrl(ctrl_buf, strlen(ctrl_buf));
+	if (ret) {
+		return ret;
+	}
+	ret = do_ftp_recv_ctrl(true, FTP_CODE_331);
+	if (ret == FTP_CODE_331) {
+		/* send password if requested */
+		sprintf(ctrl_buf, CMD_PASS, password);
+		ret = do_ftp_send_ctrl(ctrl_buf, strlen(ctrl_buf));
+		if (ret) {
+			return ret;
+		}
+		ret = do_ftp_recv_ctrl(true, FTP_CODE_230);
+	}
+	if (ret != FTP_CODE_230) {
+		return ret;
+	}
+
+	client.connected = true;
+	/* Start keep alive timer */
+	if (keepalive_time > 0) {
+		k_timer_start(&keepalive_timer, K_SECONDS(1),
+			keepalive_time * MSEC_PER_SEC);
+	}
+
+	return ret;
+}
+
+/**@brief API to close FTP connection
+ */
+int ftp_close(void)
+{
+	int ret;
+
+	ret = do_ftp_send_ctrl(CMD_QUIT, sizeof(CMD_QUIT) - 1);
+	if (ret == 0) {
+		ret = do_ftp_recv_ctrl(true, FTP_CODE_221);
+		k_timer_stop(&keepalive_timer);
+	}
+
+	close(client.sock);
+	client.connected = false;
+	return ret;
+}
+
+/**@brief API to get FTP status
+ */
+int ftp_status(void)
+{
+	int ret;
+
+	/* get server system type */
+	ret = do_ftp_send_ctrl(CMD_SYST, sizeof(CMD_SYST) - 1);
+	if (ret == 0) {
+		do_ftp_recv_ctrl(true, FTP_CODE_215);
+	}
+
+	/* get server and connection status */
+	ret = do_ftp_send_ctrl(CMD_STAT, sizeof(CMD_STAT) - 1);
+	if (ret == 0) {
+		ret = do_ftp_recv_ctrl(true, FTP_CODE_211);
+		if (ret != FTP_CODE_211) {
+			do {
+				ret = do_ftp_recv_ctrl(true, FTP_CODE_211);
+				if (ret == -ETIMEDOUT || ret == FTP_CODE_211) {
+					break;
+				}
+			} while (1);
+		}
+	}
+
+	return ret;
+}
+
+/**@brief API to set FTP transfer type
+ */
+int ftp_type(enum ftp_trasfer_type type)
+{
+	int ret;
+
+	if (type == FTP_TYPE_ASCII) {
+		ret = do_ftp_send_ctrl(CMD_TYPE_A, sizeof(CMD_TYPE_A) - 1);
+	} else if (type == FTP_TYPE_BINARY) {
+		ret = do_ftp_send_ctrl(CMD_TYPE_I, sizeof(CMD_TYPE_I) - 1);
+	} else {
+		return -EINVAL;
+	}
+	if (ret == 0) {
+		ret = do_ftp_recv_ctrl(true, FTP_CODE_200);
+	}
+
+	return ret;
+}
+
+/**@brief API to keep alive of FTP connection
+ */
+int ftp_keepalive(void)
+{
+	int ret;
+
+	ret = do_ftp_send_ctrl(CMD_NOOP, sizeof(CMD_NOOP) - 1);
+	if (ret == 0) {
+		ret = do_ftp_recv_ctrl(true, FTP_CODE_200);
+	}
+
+	return ret;
+}
+
+/**@brief API to print working directory
+ */
+int ftp_pwd(void)
+{
+	int ret;
+
+	ret = do_ftp_send_ctrl(CMD_PWD, sizeof(CMD_PWD) - 1);
+	if (ret == 0) {
+		ret = do_ftp_recv_ctrl(true, FTP_CODE_257);
+	}
+
+	return ret;
+}
+
+/**@brief API to list information of folder or file
+ */
+int ftp_list(const char *options, const char *target)
+{
+	int ret;
+	char list_cmd[128];
+
+	/* Always set Passive mode for nRF9160 to act as TCP client */
```

Comment:
I don't think we should reference any specific SoCs here, the lib should be SoC agnostic.

Reply:
It's due to fact that, Active mode will request nRF91 side to act as TCP Server, which we would like to avoid. So always Passive mode, i.e. for data channel remote FTP Server is still TCP server. This is reflected in establish_data_channel().

Rephrased this line a bit.